drivers/disk/flashdisk.c

```diff
@@ -297,31 +329,41 @@ static int disk_flash_access_write(struct disk_info *disk, const uint8_t *buff,
 
 	/* remaining partial block */
 	if (remaining) {
-		if (update_flash_block(disk, fl_addr, remaining, buff) < 0) {
-			return -EIO;
+		if (flashdisk_cache_write(ctx, fl_addr, remaining, buff) < 0) {
+			rc = -EIO;
+			goto end;
 		}
 	}
 
+end:
+	k_mutex_unlock(&ctx->lock);
+
 	return 0;
 }
 
 static int disk_flash_access_ioctl(struct disk_info *disk, uint8_t cmd, void *buff)
 {
+	int rc;
 	struct flashdisk_data *ctx;
 
 	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
 
 	switch (cmd) {
 	case DISK_IOCTL_CTRL_SYNC:
-		return 0;
+		k_mutex_lock(&ctx->lock, K_FOREVER);
```

Comment:
Move mutex lock/unlock into flashdisk_cache_commit.

Reply:
I don't agree. Zephyr mutex is recursive so acquiring already acquired mutex would not cause issues, but the cache functions are really supposed to be called with the mutex held. Note that `flashdisk_cache_commit()` can also be called from `disk_flash_access_write()` with the lock being held.

Reply:
Ok, it is also possible to create a flashdisk_ctrl_sync routine that takes the locks, calls cache_commit and unlocks. But you can leave it as is.